src/transformers/models/mistral/modeling_tf_mistral.py

```diff
@@ -0,0 +1,1068 @@
+        mask = tf.tile(mask, [bsz, 1, 1, 1])
+    else:
+        # When batch size is static, directly use broadcast_to
+        mask = tf.broadcast_to(mask[None, None, :, :], (bsz, 1, tgt_len, tgt_len + past_key_values_length))
+
+    return mask
+
+
+def _expand_mask(mask, dtype, tgt_len=None):
+    """
+    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
+    """
+    bsz, src_len = shape_list(mask)
+    tgt_len = tgt_len if tgt_len is not None else src_len
+
+    expanded_mask = tf.expand_dims(tf.expand_dims(mask, 1), 1)
+    expanded_mask = tf.broadcast_to(expanded_mask, [bsz, 1, tgt_len, src_len])
+
+    inverted_mask = 1.0 - tf.cast(expanded_mask, dtype)
+
+    return tf.where(
+        tf.cast(inverted_mask, bool), tf.fill(dims=shape_list(inverted_mask), value=tf.float32.min), inverted_mask
+    )
+
+
+class TFMistralRMSNorm(keras.layers.Layer):
+    def __init__(self, hidden_size, eps=1e-6, **kwargs):
+        """
+        TFMistralRMSNorm is equivalent to T5LayerNorm
+        """
+        super().__init__(**kwargs)
+        self.hidden_size = hidden_size
+        self.variance_epsilon = eps
+
+    def build(self, input_shape=None):
+        self.weight = self.add_weight(
+            name="weight",
+            shape=self.hidden_size,
+            initializer="ones",
+        )
+        if self.built:
+            return
+        self.built = True
+
+    def call(self, hidden_states):
+        input_dtype = hidden_states.dtype
+        hidden_states = tf.cast(hidden_states, tf.float32)
+        variance = tf.reduce_mean(tf.square(hidden_states), axis=-1, keepdims=True)
+        hidden_states = tf.divide(hidden_states, tf.sqrt(variance + self.variance_epsilon))
+        return self.weight * tf.cast(hidden_states, input_dtype)
+
+
+# Verification: https://colab.research.google.com/gist/ariG23498/f8d8131b795a131b93d99e70ee93c192/scratchpad.ipynb
+class TFMistralRotaryEmbedding(keras.layers.Layer):
+    def __init__(self, dim, max_position_embeddings=2048, base=10000, **kwargs):
+        super().__init__(**kwargs)
+
+        self.dim = dim
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        self.inv_freq = 1.0 / (self.base ** (tf.range(start=0, limit=self.dim, delta=2, dtype=tf.float32) / self.dim))
+        self._set_cos_sin_cache(seq_len=max_position_embeddings, dtype=tf.float32)
+
+    def _set_cos_sin_cache(self, seq_len, dtype):
+        self.max_seq_len_cached = seq_len
+        t = tf.cast(tf.range(self.max_seq_len_cached, dtype=tf.int64), self.inv_freq.dtype)
+
+        # freqs = tf.tensordot(t, self.inv_freq, axes=0)
+        freqs = tf.einsum("i,j->ij", t, self.inv_freq)
+        # Different from paper, but it uses a different permutation in order to obtain the same calculation
+        emb = tf.concat([freqs, freqs], axis=-1)
+
+        self.cos_cached = tf.cast(tf.cos(emb), dtype)
+        self.sin_cached = tf.cast(tf.sin(emb), dtype)
```

Comment:
In general, caching attributes like this in the forward pass doesn't work well in TF, because the forward pass is often compiled. However, in TF we don't need to do it anyway, because compilation takes care of a lot of these details for us! Check out the `RotaryEmbedding` class in `TFEsm` or `TFGemma`, and again, you can copy it if it's helpful.

Reply:
I think the simplest solution is actually to get rid of `cos_cached` and `sin_cached` entirely, and just compute `t`, `freqs`, `emb`, `cos_values` and `sin_values` in `call()`. In other words, remove `set_cos_sin_cache()` entirely, copy the relevant code into `call()`, and delete the lines where you set `self.cos_cached` and `self.sin_cached`. 

XLA will figure out which of these are constant and can be cached by itself, so there should be no performance penalty!

Reply:
I have updated the implementation. Is that what you were looking for?